rust/lance/src/dataset.rs

```diff
@@ -3117,6 +3118,111 @@ mod tests {
         dataset.validate().await.unwrap();
     }
 
+    #[rstest]
+    #[tokio::test]
+    async fn test_merge_on_row_id(
+        #[values(LanceFileVersion::Legacy, LanceFileVersion::Stable)]
+        data_storage_version: LanceFileVersion,
+        #[values(false, true)] use_stable_row_id: bool,
+    ) {
+        // Tests a merge that spans multiple batches within files
+
+        // This test also tests "null filling" when merging (e.g. when keys do not match
+        // we need to insert nulls)
+
+        let data = lance_datagen::gen()
+            .col("key", array::step::<Int32Type>())
+            .col("value", array::fill_utf8("value".to_string()))
+            .into_reader_rows(RowCount::from(1_000), BatchCount::from(10));
+
+        let test_dir = tempdir().unwrap();
+        let test_uri = test_dir.path().to_str().unwrap();
+
+        let write_params = WriteParams {
+            mode: WriteMode::Append,
+            data_storage_version: Some(data_storage_version),
+            max_rows_per_file: 1024,
+            max_rows_per_group: 150,
+            enable_move_stable_row_ids: use_stable_row_id,
+            ..Default::default()
+        };
+        Dataset::write(data, test_uri, Some(write_params.clone()))
+            .await
+            .unwrap();
+
+        let mut dataset = Dataset::open(test_uri).await.unwrap();
+        assert_eq!(dataset.fragments().len(), 10);
+        assert_eq!(dataset.manifest.max_fragment_id(), Some(9));
+
+        let data = dataset.scan().with_row_id().try_into_batch().await.unwrap();
+        let row_ids = data.column_by_name(ROW_ID).unwrap();
+        let new_schema = Arc::new(ArrowSchema::new(vec![
+            ArrowField::new("rowid", DataType::UInt64, false),
+            ArrowField::new("new_value", DataType::UInt64, false),
+        ]));
+        let new_batch =
+            RecordBatch::try_new(new_schema.clone(), vec![row_ids.clone(), row_ids.clone()])
+                .unwrap();
+        let new_data = RecordBatchIterator::new(vec![Ok(new_batch)], new_schema.clone());
+        dataset.merge(new_data, ROW_ID, "rowid").await.unwrap();
+        dataset.validate().await.unwrap();
```

Comment:
I'd like us to assert a few more things in this test:

1. `dataset` has the expected final schema `key, value, new_value`.
2. The values are what we expect. For this, you should avoid using the same values in each column. Otherwise, the test could pass even if there is a bug that uses the wrong column's values. Right now, you use `row_ids.clone()` for both `rowid` and `new_value`.
3. This works even if you shuffle the data. I would recommend using [take_record_batch()](https://docs.rs/arrow/latest/arrow/compute/fn.take_record_batch.html) to reorder the `new_batch` so the row ids are out-of-order.